make/gotools.mk

```diff
@@ -0,0 +1,126 @@
+#   GOTOOLS_BIN: the directory in which binaries are stored; defaults to $(GOTOOLS_ROOT)/bin. This directory is added
+#                to PATH.
+#
+# This file defines a single (user-facing) macro, `go-tool`, which can be invoked via
+#   $(call go-tool VARNAME, go-pkg [, module-root])
+# where go-pkg can be:
+# - an absolute Go import path with an explicit version, e.g.,
+#     github.com/golangci/golangci-lint/cmd/golangci-lint@v1.49.0. In this case, the tool is installed via `go install`,
+#     and module information from the local workspace is ignored, in accordance with the normal behavior of go install
+#     with an explicit version given.
+# - an absolut Go import path WITHOUT a version, e.g., github.com/golangci/golangci-lint/cmd/golangci-lint. In this
+#     case, the tool is installed via `go install` from the module rooted at $(GOTOOLS_PROJECT_ROOT), or, if
+#     module-root is given, from the module rooted at that (relative) path. I.e., go-pkg must be a provided by a module
+#     listed as a requirement in <module-root>/go.mod.
+# - a relative Go import path (WITHOUT a version), e.g., ./tools/roxvet. In this case, the tool is installed via
+#     `go install` from the module rooted at $(GOTOOLS_PROJECT_ROOT).
+#
+# Invoking go-tool will set up Makefile rules to build the tools, using reasonable strategies for caching to avoid
+# building a tool multiple times. In particular:
+# - when using an absolute Go import path with a version, the rule is set up such that the `go install` command is only
+#   run once.
+# - when using an absolute Go import path without a version, the rule is set up such that the `go install` command is
+#   re-run only when the respective go.mod file changes.
+# - when using a relative Go import path, the rule is set up such that the `go install` command is re-run on every
+#   `make` invocation.
+# Note that `go install` uses a pretty effective caching strategy under the hood, so even with relative import path,
+# you should not expect noticeable latency.
+#
+# In addition to setting up rules, invoking go-tool will also set the value of the variable `VARNAME` to the (canonical)
+# location of the respective tool's binary, which is $(GOTOOLS_BIN)/<binary basename>. `$(VARNAME)` should be used as
+# the only way of both invoking the tool in the Makefile as well as expressing a dependency on the installation of the
+# tool.
+# In contexts where `$(VARNAME)` is not available, such as when invoking external shell scripts from the Makefile, the
+# tools can also be accessed by their basename without a path prefix, as `$(GOTOOLS_BIN)` is added to the `PATH` when
+# including this file.
+#
+# This file also defines two static, global targets:
+#   gotools-clean: this removes all gotools-related data
+#   gotools-all: this builds all gotools.
+
+GOTOOLS_PROJECT_ROOT ?= $(CURDIR)
+GOTOOLS_ROOT ?= $(GOTOOLS_PROJECT_ROOT)/.gotools
+GOTOOLS_BIN ?= $(GOTOOLS_ROOT)/bin
+
+PATH := $(GOTOOLS_BIN):$(PATH)
+export PATH
```

Comment:
Does this have any effect on the parent `Makefile`? I suppose that's the way due to which `generate-junit-reports.sh` works when called from the `Makefile`. This seems too sneaky, implicit and prone to differences between user-default environment and the environment constructed from the `Makefile`.

It seems, that you're constructing something similar to Python's VirtualEnv here. VirtualEnv has `activate` script which developers source in shell, and get all local binaries available for interactive use.
Maybe we don't need to be that fancy and just hardcode binaries to be called from `.gotools/` directory. If the binary is there - the script would use it. If the binary isn't there, the developer needs to get it through invoking `Makefile`.

Reply:
> Does this have any effect on the parent `Makefile`?

TTBOMK, yes, `include` in `Makefiles` are just inserted inline.

> This seems too sneaky, implicit and prone to differences between user-default environment and the environment constructed from the `Makefile`.

Is this about this happening in an `include`d file, or about the `PATH` being modified in the `Makefile`? It sounds like the latter, and I don't think I agree. "User-default environment" is an anti-pattern when it comes to building software, the build logic has every right to shield itself from the user-default environment, and that necessarily means deviating from it. One of the points of using a `Makefile` is that it provides a uniform way and entry-point to build commands. If somebody can't just copy something out of the `Makefile` into their terminal without observing different behavior, well, tough luck, the `Makefile` is there for a reason (yes, there are cases when you want to run stuff directly in the terminal, but those are corner-cases).

In the concrete case, I don't see any evidence that `generate-junit-reports.sh` is intended to be invoked outside of the respective `Makefile` rule, and just because it's there doesn't mean you can just run it and expect it to work. Is your concern that it's easy to run this script and get the _wrong_ result (using a different version) without noticing? Would you be okay with instead `export` the `XYZ_BIN` variables, and do a `[[ -x "$GO_JUNIT_REPORT_BIN" ]] || die "..."` in the script? I don't think there's a big difference in the two approaches, in both cases we rely on environment set up by the `Makefile`. The one advantage I can see is that we now _reserve_ the invocation by basename only for when we depend on the user-default environment, so it makes things somewhat less ambiguous.

But there will always be cases when we can't do this, for example, if we invoke a tool that we don't control that depends on the `PATH`. We invoke our protobuf compilation with `protoc ... --gofast_out=...`. This causes `protoc` to look for a `protoc-gen-gofast` binary in the path (generally, `--{x}_out` causes `protoc-gen-{x}` to be used). Here, it seems a requirement to set a custom `PATH`.

Reply:
With regards to `generate-junit-reports.sh` file, I suggest to simply inline its content in the `Makefile`. That would be correct according to the logic that the `Makefile` should be the entry point that you propose. There `$GO_JUNIT_REPORT_BIN` can be safely used instead of `go-junit-report`.

Reply:
Yeah, good point about remove `generate-junit-reports.sh`. I've removed it and also made the respective rules more `make`-y`.

Good point about `PATH` being potentially unneeded, especially after the above change. I have removed it now, let's see if things break.

We still have the reference to the tool in the `mockgen-wrapper` script. Do you have an opinion on the `which-<tool>` target I sketched in the conversation with Marcin above?

Reply:
Thanks.
Commented on `mockgen-wrapper` thread.
Resolving this one.